Description: adjust logic to report on why a file is needed in verbose mode.
Author: Kees Cook <kees@ubuntu.com>
Bug: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=455082

Index: debmirror-2.4.4/debmirror
===================================================================
--- debmirror-2.4.4.orig/debmirror	2010-04-05 12:21:29.255149875 -0700
+++ debmirror-2.4.4/debmirror	2010-04-05 12:21:44.362648442 -0700
@@ -1436,19 +1436,29 @@
 # It will return true if the tests show the file matches.
 sub check_file {
   my ($filename, $size, $md5sum)=@_;
-  if (-f $filename and ($size == -s _ || $size == -1)) {
+  if (! -f $filename) {
+    say("Missing: $filename") if ($verbose);
+    return 0;
+  }
+  my $disksize = -s _;
+  if ($size == $disksize || $size == -1) {
     if ($check_md5sums || $size == -1) {
       open HANDLE, $filename or
 	die "$filename: $!";
       $md5->addfile(*HANDLE);
       my $digest = $md5->hexdigest;
-      return ($md5sum eq $digest);
+      return 1 if ($md5sum eq $digest);
+      say(sprintf("Mismatch '$filename': sum is %s, expected %s",$digest,$md5sum))
+        if ($verbose);
     }
     else {
       # Assume it is ok, w/o md5 check.
       return 1;
     }
   }
+  elsif ($verbose) {
+    say(sprintf("Mismatch '$filename': size is %d, expected %d",$disksize,$size));
+  }
   return 0;
 }
 
